From eaec76d04d53c4b03008c33d49ab06f2e974cc42 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Mon, 22 Dec 2025 03:07:23 +0100
Subject: [PATCH 98/98] net: phy: icplus: Add support for managing the clocks
 on IP101A/G - WiP

TODO: commit message

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

[dts: rebased onto 6.6.144 for 6.6-lts - devm_clk_get_optional_enabled_with_rate()
 doesn't exist at this baseline; kept the split
 devm_clk_get_optional_enabled() + clk_set_rate(), same as done for
 drivers/net/phy/smsc.c]
---
 drivers/net/phy/icplus.c | 51 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/drivers/net/phy/icplus.c b/drivers/net/phy/icplus.c
index c0c4f19cfb6a00..b8ee9ce07b1d47 100644
--- a/drivers/net/phy/icplus.c
+++ b/drivers/net/phy/icplus.c
@@ -6,6 +6,8 @@
  */
 #include <linux/kernel.h>
 #include <linux/string.h>
+#include <linux/clk.h>
+#include <linux/clk-provider.h>
 #include <linux/errno.h>
 #include <linux/unistd.h>
 #include <linux/interrupt.h>
@@ -201,10 +203,36 @@ static int ip175c_config_aneg(struct phy_device *phydev)
 	return 0;
 }
 
+static int ip101a_g_register_rmii_ref_clk(struct device *dev)
+{
+	struct clk_hw *rmii_ref_clk;
+	char *name;
+
+	if (!IS_ENABLED(CONFIG_COMMON_CLK))
+		return 0;
+
+	name = devm_kasprintf(dev, GFP_KERNEL, "%s-c50m_o", dev_name(dev));
+	if (!name)
+		return -ENOMEM;
+
+	rmii_ref_clk = devm_clk_hw_register_fixed_factor_index(dev, name, 0, 0,
+							       2, 1);
+	if (IS_ERR(rmii_ref_clk))
+		return dev_err_probe(dev, PTR_ERR(rmii_ref_clk),
+				     "Failed to register RMII REF_CLK clock output\n");
+
+	return devm_of_clk_add_hw_provider(dev, of_clk_hw_simple_get,
+					   rmii_ref_clk);
+}
+
 static int ip101a_g_probe(struct phy_device *phydev)
 {
 	struct device *dev = &phydev->mdio.dev;
 	struct ip101a_g_phy_priv *priv;
+	unsigned long in_clk_mhz;
+	bool has_ref_clk_input;
+	struct clk *clk;
+	int ret;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -229,6 +257,29 @@ static int ip101a_g_probe(struct phy_device *phydev)
 
 	phydev->priv = priv;
 
+	has_ref_clk_input = device_property_match_string(dev, "clock-names",
+							 "ref_clk") >= 0;
+	if (has_ref_clk_input)
+		in_clk_mhz = 50;
+	else
+		in_clk_mhz = 25;
+
+	clk = devm_clk_get_optional_enabled(dev, NULL);
+	if (IS_ERR(clk))
+		return dev_err_probe(dev, PTR_ERR(clk),
+				     "Failed to get %luMHz clock input\n",
+				     in_clk_mhz);
+
+	ret = clk_set_rate(clk, in_clk_mhz * 1000 * 1000);
+	if (ret)
+		return ret;
+
+	if (phydev->interface == PHY_INTERFACE_MODE_RMII && !has_ref_clk_input) {
+		ret = ip101a_g_register_rmii_ref_clk(dev);
+		if (ret)
+			return ret;
+	}
+
 	return 0;
 }
 
